src/DistributedComputations/transposable_field.jl

```diff
@@ -0,0 +1,166 @@
+using Oceananigans.Grids: architecture
+using Oceananigans.Architectures: on_architecture
+
+struct TransposableField{FX, FY, FZ, YZ, XY, C, Comms}
+    xfield :: FX # X-direction is free
+    yfield :: FY # Y-direction is free
+    zfield :: FZ # Z-direction is free (original field)
+    yzbuff :: YZ # if `nothing` slab decomposition with `Ry == 1`
+    xybuff :: XY # if `nothing` slab decomposition with `Rx == 1`
+    counts :: C
+    comms  :: Comms
+end
+
+const SlabYFields = TransposableField{<:Any, <:Any, <:Any, <:Nothing} # Y-direction is free
+const SlabXFields = TransposableField{<:Any, <:Any, <:Any, <:Any, <:Nothing} # X-direction is free
+
+"""
+    TransposableField(field_in, FT = eltype(field_in); with_halos = false)
+
+Constructs a TransposableField object that containes the allocated memory and the ruleset required
+for distributed transpositions.
+
+# Arguments
+- `field_in`: The input field. It needs to be in a _z-free_ configuration (i.e. ranks[3] == 1).
+- `FT`: The element type of the field. Defaults to the element type of `field_in`.
+- `with_halos`: A boolean indicating whether to include halos in the field. Defaults to `false`.
+"""
+function TransposableField(field_in, FT = eltype(field_in); with_halos = false)
```

Comment:
This needs more documentation to explain how it works

Reply:
Flagging again that there are some unresolved comments in this file.

Reply:
I have added some documentation to this type.